cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov-node/sriov_node.sh

```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+
+SCRIPT_PATH=${SCRIPT_PATH:-$(dirname "$(realpath "$0")")}
+
+CONFIGURE_VFS_SCRIPT_PATH="${SCRIPT_PATH}/configure_vfs.sh"
+
+function sriov_node::discover_host_pfs() {
+  local -r sriov_pfs=( $(find /sys/class/net/*/device/sriov_numvfs) )
+  [ "${#sriov_pfs[@]}" -eq 0 ] && echo "FATAL: Could not find available sriov PFs on host" >&2 && return 1
+
+  local pf_name
+  local pf_names=()
+  for pf in "${sriov_pfs[@]}"; do
+    pf_name="${pf%%/device/*}"
+    pf_name="${pf_name##*/}"
+    if [ $(echo "${PF_BLACKLIST[@]}" | grep "${pf_name}") ]; then
+      continue
+    fi
+
+    pfs_names+=( $pf_name )
+  done
+
+  echo "${pfs_names[@]}"
+}
+
+function sriov_node::configure_sriov_pfs_and_vfs() {
+  local -r nodes_array=($1)
+  local -r pfs_names_array=($2)
+  local -r pf_count_per_node=$3
+
+  local -r config_vf_script=$(basename "$CONFIGURE_VFS_SCRIPT_PATH")
+  local pfs_to_move=()
+  local pfs_array_offset=0
+  local node_exec
+
+  # 'iplink' learns which network namespaces there are by checking /var/run/netns
+  mkdir -p /var/run/netns
+  for node in "${nodes_array[@]}"; do
+    prepare_node_netns "$node"
+
+    ## Move PF's to node netns
+    # Slice '$pfs_names_array' to have unique silce for each node
+    # with '$pf_count_per_node' PF's names
+    pfs_to_move=( "${pfs_names_array[@]:$pfs_array_offset:$pf_count_per_node}" )
+    echo "Moving '${pfs_to_move[*]}' PF's to '$node' netns"
+    for pf_name in "${pfs_to_move[@]}"; do
+      move_pf_to_node_netns "$node" "$pf_name"
+    done
+    # Increment the offset for next slice
+    pfs_array_offset=$((pfs_array_offset + pf_count_per_node))
+
+    # KIND mounts sysfs as read-only by default, remount as R/W"
+    node_exec="docker exec $node"
+    $node_exec mount -o remount,rw /sys
+    $node_exec chmod 666 /dev/vfio/vfio
+
+    # Create and configure SRIOV Virtual Functions on SRIOV node
+    docker cp "$CONFIGURE_VFS_SCRIPT_PATH" "$node:/"
+    $node_exec bash -c "DRIVER=$VFS_DRIVER DRIVER_KMODULE=$VFS_DRIVER_KMODULE ./$config_vf_script"
+
+    _kubectl label node $node $SRIOV_NODE_LABEL
+  done
+}
+
+function prepare_node_netns() {
+  local -r node_name=$1
+  local -r node_pid=$(docker inspect -f '{{.State.Pid}}' "$node_name")
+
+  # Docker does not create the required symlink for a container netns
+  # it perverts iplink from learning that container netns.
+  # Thus it is necessary to create symlink between the current
+  # worker node (container) netns to /var/run/netns (consumed by iplink)
+  # Now the node container netns named with the node name will be visible.
+  ln -sf "/proc/$node_pid/ns/net" "/var/run/netns/$node_name"
+}
+
+function move_pf_to_node_netns() {
+  local -r node_name=$1
+  local -r pf_name=$2
+
+  # Move PF to node network-namespace
+  ip link set "$pf_name" netns "$node_name"
+  # Ensure current PF is up
+  ip netns exec "$node_name" ip link set up dev "$pf_name"
```

Comment:
It used to be 
https://github.com/kubevirt/kubevirtci/pull/570/files#diff-499cff045d8a875e8b39f4a9fd59354eb99e8f8d91a95ebc70fbf42d2d5b5beaL51

Now it means that you won't be able to spin two clusters locally at the very same time,
you will still be able to spin them if you spin one, wait for its to done spinning and then spin the 2nd.
So either please fix it, or if we don't want parallel spinning (just parallel running) locally, please update about it in the doc.

(this comment affects only locally)

Reply:
Why it wont affect CI jobs?

Reply:
I updated the doc.

Reply:
It won't affect CI because the undercloud CNI allocates orthogonal PFs to each job,
the race that could happen is just if the job itself runs two clusters at the very same time,
this what can happen locally, because locally we don't have the 2nd layer of job separation.

Thanks

Reply:
last paragraph of https://github.com/kubevirt/kubevirtci/pull/570#discussion_r609365714 explained

see please the elaboration
https://github.com/kubevirt/kubevirtci/pull/570#discussion_r609497314

Reply:
> last paragraph of [#570 (comment)](https://github.com/kubevirt/kubevirtci/pull/570#discussion_r609365714) explained
> 
You comment on a function not on the right place we dont have context and its confusing.
Please remove it from there and write it here.
If you think you saw something that need to be change, and changing it means to touch multiple places
comment about it on the PR page instead.

Reply:
Actually its not a mistake, its this code 
https://github.com/kubevirt/kubevirtci/pull/570/files/73fa656af60f4b386531306e53d54249840854a9#diff-0ab4a7b3ef24d741b22df5cd4fc6fd14ac398ecd2e6b831aa400ba7a7bd3494eR82-R84

that possible has consequences on those lines,
it may flake now, because it takes time to move the PF to the container
All the details are here https://github.com/kubevirt/kubevirtci/pull/570/files/73fa656af60f4b386531306e53d54249840854a9#r609497314
but i will try to focus and make it straight to the pint below in this comment:

I can't delete it now and order it, because there are discussions already.

The main issue is that removing the code
https://github.com/kubevirt/kubevirtci/pull/570/files#diff-499cff045d8a875e8b39f4a9fd59354eb99e8f8d91a95ebc70fbf42d2d5b5beaL51

caused at least two things:
1. We can't spin local clusters parallell in the very same time, Edy find it important to preserve it.
2. `ip netns exec "$node_name" ip link set up dev "$pf_name"` might be flaky now, as moving the PF is async, it might continue `ip link set` even that the PF is not moved yet by the previous command, but it existed aready.

I would start focusing on those, imo it can be even be on future PRs, upon need.
More possible deeper suggestions in the reference, but we should start with those 2 i think.

Reply:
1. Kubevirt  changes frequently as you may know, and recently the migration code also changes and we dont have SRIOV migration tested on CI just yet which is very bad. 
We already had to prepare a patch after there was a change on Kubevirt that broke SRIOV migration (converting libvirt domains to have persistent config, you can have more details on https://github.com/kubevirt/kubevirt/pull/5152 https://github.com/kubevirt/kubevirt/pull/5150)
Due to this urgency I prefer to handle this after this PR get into Kubevirt so SRIOV migration will be tested on CI.

If we will agree that this is something we can leave without it, we dont need to followup on this until there will be a demand for it.
I am saying this because AFAIU the main purpose of it is to optimize CI and we dont know of someone  that uses it for development.
  
Have that said its still possible to run two cluster on a local setup its not totally not functional, 
I added this constrain that creating two clusters should be sequential to the README.

2. Personally I never encountered problems with it, I tested it many times on CI and even more on local setup.
Lets handles this when we see it happen like you suggests.

Reply:
Fine by me, but see note please, just before we decide that.

Note: Returning the code 
https://github.com/kubevirt/kubevirtci/pull/570/files#diff-499cff045d8a875e8b39f4a9fd59354eb99e8f8d91a95ebc70fbf42d2d5b5beaL51
Would fix both imo, without nothing else needed.
```
    if ip link set "$pf_name" netns "$node"; then
      if timeout 10s bash -c "until ip netns exec $node ip link show $pf_name > /dev/null; do sleep 1; done"; then
        pf_array+=("$pf_name")
        [ "${#pf_array[@]}" -eq "$pf_count_per_node" ] && break
      fi
    fi
```

It verify that the PF is moved, and allow parallel spin up.
So it solves both [1] and [2].

The only corner case which is super rare if even happen (didn't saw), is that wasn't move in 10 seconds but it was moved in the background later, and then the nodes will have a PF that wasn't meant to be there.
This why I suggested an assert on the expected PF / VF, but again, we don't need it in this PR, nor to handle the theoretic corner case IMO.

Btw, did you saw a case where you needed 
`ip netns exec "$node_name" ip link set up dev "$pf_name"`?
I didn't personally, but anyhow its a good barrier in case we don't solve [2]

Reply:
Fortunately adding warping `ip link set <pf> netns <node>` with `timeout` and `until` wont enable creating clusters in parallel.
The main problem is the fact that we pick PF's before moving them to a node netns in advance, 
in case of a failure we cant pick the next PF, like on the current implementation (which is pretty straightforward because it supports only one node).

The function `node::configure_sriov_pfs_and_vfs` iterates through the nodes and move the PF's for each node netns.
Each iteration (for each node) PF's names array is sliced according to `PF_COUNT_PER_NODE` flag.
Then each PF on that slice is moved to the current node netns.
In case moving PF to node netns is failed for any reason (`ip link set <pf> netns <node>`) 
cluster-up will stop with an error.

Considering the scenario we create two clusters in parallel, `cluster1` and `cluster2`, where `cluster1` started first:
both `cluster2` and `cluster1` picked  the same PF's to move, 
cluster-up would fail on moving PF command by `cluster2` with an error that the PF does not exists (on the default netns).
We could do best effort and not fail on this, but we would end up with non consistent setups where not all nodes has the desired amount of PF's which is bad idea.

Due to the reasons I mentioned on this thread I prefer to address the problem I described on a follow up PR in case we need it.

About warping `ip link set <pf> netns <node>` with `until` `timeout`, you mentioned that this operation is non blocking.
From what I know and saw `iplinks` operations are blocking.
Could you please elaborate or give a reference that tells that this is operation is not blocking?

Reply:
About the first discussion:
Changing the flow such that it will first move the needed PFs, with the option to `continue` and then look in the PFs it have already would solve that (instead of discover_host_pfs and then move, I already said in the past that imo its not needed to split it).
But ok don't address it on this PR.

> About warping ip link set <pf> netns <node> with until timeout, you mentioned that this operation is non blocking.
From what I know and saw iplinks operations are blocking.
Could you please elaborate or give a reference that tells that this is operation is non blocking?

Based on something Edy told me in the past, and i saw it sometimes happen, for example during developing the underlcloud CNI (unless we find out it was due to other issues, but adding this retry indeed fixed it)
I don't have in hand documentation, and since we are not handling it right now its not crucial to find one imo.